plugins/snapd/fu-snapd-plugin.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2024 Maciej Borzecki <maciej.borzecki@canonical.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <time.h>
+
+#include "fwupd-enums.h"
+#include "fwupd-error.h"
+
+#include "fu-firmware.h"
+#include "fu-progress.h"
+#include "fu-secure-boot-device.h"
+#include "fu-snapd-error.h"
+#include "fu-snapd-observer.h"
+#include "fu-snapd-plugin.h"
+#include "fu-snapd-snap.h"
+#include "glib.h"
+
+struct _FuSnapdPlugin {
+	FuPlugin parent_instance;
+
+	FuSnapdObserver *snapd_observer;
+};
+
+G_DEFINE_TYPE(FuSnapdPlugin, fu_snapd_plugin, FU_TYPE_PLUGIN)
+
+static gboolean
+fu_snapd_plugin_notify_init(FuSnapdPlugin *self, GError **error);
+
+static void
+fu_snapd_plugin_init(FuSnapdPlugin *self)
+{
+}
+
+static void
+fu_snapd_plugin_finalize(GObject *object)
+{
+	FuSnapdPlugin *self = FU_SNAPD_PLUGIN(object);
+	if (self->snapd_observer != NULL) {
+		g_object_unref(self->snapd_observer);
+		self->snapd_observer = NULL;
+	}
+}
+
+static gboolean
+fu_snapd_notify_secure_boot_dbx_write(FuSecureBootDevice *dev,
+				      FuFirmware *fw,
+				      gpointer user_data,
+				      GError **error);
+
+static gboolean
+fu_snapd_plugin_composite_prepare(FuPlugin *plugin, GPtrArray *devices, GError **error)
+{
+	FuSnapdPlugin *self = FU_SNAPD_PLUGIN(plugin);
+
+	g_debug("composite prepare");
+
+	g_return_val_if_fail(self->snapd_observer != NULL, TRUE);
+
+	/* TODO which devices are part of the update */
+	/* TODO if found UEFI DBX device, then register a callback */
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *dev = g_ptr_array_index(devices, i);
+		if (FU_IS_SECURE_BOOT_DEVICE(dev)) {
+			FuSecureBootDevice *sbdev = FU_SECURE_BOOT_DEVICE(dev);
+			FuSecureBootDeviceKind kind = fu_secure_boot_device_get_kind(sbdev);
+			if (kind != FU_SECURE_BOOT_DEVICE_KIND_UEFI_DBX) {
+				continue;
+			}
+
+			g_debug("found DBX device");
+			if (!fu_secure_boot_device_set_firmware_write_observe(
+				sbdev,
+				fu_snapd_notify_secure_boot_dbx_write,
+				(gpointer)plugin)) {
+				g_warning("cannot install firmware write observer");
+				return FALSE;
+			}
```

Comment:
there's a few layers of indirection here, but I _think_ you are forgetting to set `error` here.